File: scripts/cleanup_versions.py
import json
import urllib.request
import urllib.parse

PB_URL = 'http://115.29.162.18:8090'
ADMIN_EMAIL = 'rocky.hk@gmail.com'
ADMIN_PASSWORD = 'gz203799'

def make_request(url, method='GET', data=None, headers=None, params=None):
    if params:
        url = f"{url}?{urllib.parse.urlencode(params)}"
    
    req = urllib.request.Request(url, method=method)
    if headers:
        for k, v in headers.items():
            req.add_header(k, v)
    
    if data:
        json_data = json.dumps(data).encode('utf-8')
        req.add_header('Content-Type', 'application/json')
        req.data = json_data

    try:
        with urllib.request.urlopen(req) as response:
            return response.status, json.loads(response.read().decode('utf-8'))
    except urllib.error.HTTPError as e:
        body = e.read().decode('utf-8')
        try:
            return e.code, json.loads(body)
        except:
            return e.code, body
    except Exception as e:
        return 0, str(e)

def cleanup():
    # 1. Login as Superuser (v0.23+ style)
    print(f"🚀 Connecting to {PB_URL} and logging in as superuser...")
    status, auth_data = make_request(f"{PB_URL}/api/collections/_superusers/auth-with-password", method='POST', data={
        "identity": ADMIN_EMAIL,
        "password": ADMIN_PASSWORD
    })
    
    if status != 200:
        print(f"❌ Login failed: {auth_data}")
        return
    
    token = auth_data['token']
    headers = {"Authorization": token}
    print("✅ Login successful!")

    # 2. Fetch all software versions
    print("📦 Fetching all software versions...")
    status, versions_data = make_request(f"{PB_URL}/api/collections/software_versions/records", 
                                       params={"perPage": 500}, 
                                       headers=headers)
    
    if status != 200:
        print(f"❌ Failed to fetch versions: {versions_data}")
        return
    
    all_versions = versions_data.get('items', [])
    print(f"📊 Total version records: {len(all_versions)}")

    # 3. Group versions by brand and versionString
    groups = {}
    for v in all_versions:
        brand = v.get('brand')
        vs = (v.get('versionString') or '').strip()
        key = f"{brand}_{vs}"
        if key not in groups:
            groups[key] = []
        groups[key].append(v)

    # 4. Process duplicates
    for key, records in groups.items():
        if len(records) <= 1:
            continue
        
        print(f"\nProcessing duplicates for: [{key}] ({len(records)} records)")
        
        # Sort by created time
        records.sort(key=lambda x: x['created'])
        
        official = None
        for r in records:
            # 官方版本判断：isCustom 不是 true，且 versionString 中不包含“自定义”
            is_custom_flag = r.get('isCustom', False)
            if not is_custom_flag and '自定义' not in (r.get('versionString') or ''):
                official = r
                break
        
        if not official:
            official = records[0] # 如果都没明确标记，拿最早创建的
            
        print(f"   🏆 Selected OFFICIAL ID: {official['id']} [{official['versionString']}]")
        
        redundant_records = [r for r in records if r['id'] != official['id']]
        
        for red in redundant_records:
            red_id = red['id']
            print(f"   ⚠️ Found REDUNDANT ID: {red_id} [{red.get('versionString')}]")
            
            # 5. Migrate Users
            status, users_data = make_request(f"{PB_URL}/api/collections/users/records", 
                                           params={"filter": f'software_version_ref = "{red_id}"', "perPage": 200}, 
                                           headers=headers)
            users = users_data.get('items', []) if isinstance(users_data, dict) else []
            if users:
                print(f"      - Migrating {len(users)} users...")
                for user in users:
                    status, _ = make_request(f"{PB_URL}/api/collections/users/records/{user['id']}", 
                                           method='PATCH',
                                           data={
                                               "software_version_ref": official['id'],
                                               "software_version": official['versionString']
                                           }, 
                                           headers=headers)
                    if status != 200:
                        print(f"        ❌ User update failed for {user['id']}")

            # 6. Migrate Trips
            status, trips_data = make_request(f"{PB_URL}/api/collections/trips/records", 
                                           params={"filter": f'software_version_ref = "{red_id}"', "perPage": 500}, 
                                           headers=headers)
            trips = trips_data.get('items', []) if isinstance(trips_data, dict) else []
            if trips:
                print(f"      - Migrating {len(trips)} trips...")
                for trip in trips:
                    status, _ = make_request(f"{PB_URL}/api/collections/trips/records/{trip['id']}", 
                                           method='PATCH',
                                           data={
                                               "software_version_ref": official['id'],
                                               "software_version": official['versionString']
                                           }, 
                                           headers=headers)
                    if status != 200:
                        print(f"        ❌ Trip update failed for {trip['id']}")

            # 7. Delete Redundant Record
            print(f"      - 🗑️ Deleting redundant record: {red_id}")
            status, _ = make_request(f"{PB_URL}/api/collections/software_versions/records/{red_id}", 
                                   method='DELETE', headers=headers)
            if status >= 400:
                print(f"        ❌ Delete failed for {red_id}")

    print("\n✨ Data cleanup and migration completed successfully!")

if __name__ == "__main__":
    cleanup()

File: scripts/cleanup_versions.js
import PocketBase from 'pocketbase';

/**
 * PocketBase 数据清洗脚本
 * 功能：将带有“用户自定义”标签的版本号迁移到对应的正式版本号，并删除冗余版本。
 * 
 * 使用方法：
 * 1. 进入 Puked_web 目录: cd Puked_web
 * 2. 安装依赖: npm install
 * 3. 设置环境变量并运行:
 *    PB_ADMIN_EMAIL=your@email.com PB_ADMIN_PASSWORD=your_password node ../scripts/cleanup_versions.js
 */

const PB_URL = 'https://pb.osglab.com';
const ADMIN_EMAIL = process.env.PB_ADMIN_EMAIL;
const ADMIN_PASSWORD = process.env.PB_ADMIN_PASSWORD;

if (!ADMIN_EMAIL || !ADMIN_PASSWORD) {
    console.error('❌ 错误: 请设置环境变量 PB_ADMIN_EMAIL 和 PB_ADMIN_PASSWORD');
    process.exit(1);
}

async function cleanup() {
    const pb = new PocketBase(PB_URL);

    try {
        console.log('🚀 正在连接 PocketBase 并登录 Admin...');
        await pb.admins.authWithPassword(ADMIN_EMAIL, ADMIN_PASSWORD);
        console.log('✅ 登录成功！');

        console.log('📦 获取所有软件版本信息...');
        const allVersions = await pb.collection('software_versions').getFullList({
            requestKey: null,
        });
        console.log(`📊 共获取到 ${allVersions.length} 条版本记录`);

        const officialMap = new Map(); // key: brandId_versionString, value: record
        const customRecords = [];

        // 第一步：对版本进行分类
        for (const v of allVersions) {
            const vs = v.versionString || '';
            // 匹配“用户自定义”或“自定义”标签
            const isCustom = vs.includes('用户自定义') || vs.includes('自定义');
            
            if (isCustom) {
                customRecords.push(v);
            } else {
                const key = `${v.brand}_${vs.trim()}`;
                // 如果存在多个相同的正式版本号，保留最早创建的一个（ID 通常反映创建顺序）
                if (!officialMap.has(key)) {
                    officialMap.set(key, v);
                }
            }
        }

        console.log(`🔍 识别到 ${customRecords.length} 条待处理的自定义版本记录`);

        // 第二步：执行迁移和清理
        for (const custom of customRecords) {
            // 提取核心版本号（移除各种格式的自定义标签）
            const coreVersion = custom.versionString
                .replace(/\s*\(?用户自定义\)?\s*/g, '')
                .replace(/\s*\(?自定义\)?\s*/g, '')
                .trim();
            
            const key = `${custom.brand}_${coreVersion}`;
            const official = officialMap.get(key);

            if (official) {
                console.log(`\n🔄 迁移: "${custom.versionString}" -> 官方版本: "${official.versionString}" (ID: ${official.id})`);

                // 迁移关联的用户
                const users = await pb.collection('users').getFullList({
                    filter: `software_version_ref = "${custom.id}"`,
                    requestKey: null,
                });
                if (users.length > 0) {
                    console.log(`   - 正在迁移 ${users.length} 个用户...`);
                    for (const user of users) {
                        await pb.collection('users').update(user.id, {
                            software_version_ref: official.id,
                            software_version: official.versionString
                        });
                    }
                }

                // 迁移关联的行程 (Trips)
                const trips = await pb.collection('trips').getFullList({
                    filter: `software_version_ref = "${custom.id}"`,
                    requestKey: null,
                });
                if (trips.length > 0) {
                    console.log(`   - 正在迁移 ${trips.length} 个行程...`);
                    for (const trip of trips) {
                        await pb.collection('trips').update(trip.id, {
                            software_version_ref: official.id,
                            software_version: official.versionString
                        });
                    }
                }

                // 删除冗余的自定义版本记录
                console.log(`   - 删除冗余记录: ${custom.id}`);
                await pb.collection('software_versions').delete(custom.id);

            } else {
                // 如果没有找到对应的官方版本，则将此记录更名为核心版本号（即“转正”）
                console.log(`\n✨ 转正: "${custom.versionString}" 无对应官方版本，更名为 "${coreVersion}"`);
                await pb.collection('software_versions').update(custom.id, {
                    versionString: coreVersion
                });
                
                // 同步更新关联数据中的显示字符串
                const users = await pb.collection('users').getFullList({
                    filter: `software_version_ref = "${custom.id}"`,
                    requestKey: null,
                });
                for (const user of users) {
                    await pb.collection('users').update(user.id, {
                        software_version: coreVersion
                    });
                }

                const trips = await pb.collection('trips').getFullList({
                    filter: `software_version_ref = "${custom.id}"`,
                    requestKey: null,
                });
                for (const trip of trips) {
                    await pb.collection('trips').update(trip.id, {
                        software_version: coreVersion
                    });
                }

                // 将该版本记录视为该品牌的“正式版”
                officialMap.set(key, { ...custom, versionString: coreVersion });
            }
        }

        console.log('\n✨ 所有数据迁移与清理任务已成功完成！');

    } catch (error) {
        console.error('\n❌ 处理过程中发生错误:', error.message || error);
        if (error.data) {
            console.error('详细信息:', JSON.stringify(error.data, null, 2));
        }
    }
}

cleanup();
